5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py

```diff
@@ -0,0 +1,245 @@
+    pathlib.Path(
+        "../0.train_logistic_regression/models/log_reg_fs_plate_4_final_downsample.joblib"
+    )
+)
+
+# Load the feature-selected QC plate 4 (ensure it includes both features and labels)
+plate_4_QC = pd.read_parquet(
+    pathlib.Path(
+        "../../3.process_cfret_features/data/single_cell_profiles/localhost231120090001_sc_feature_selected.parquet"
+    )
+)
+
+# Filter the test_data to only include rows where:
+# - Metadata_heart_number == 7 and Metadata_treatment == "DMSO"
+# - OR Metadata_heart_number == 29
+holdout_QC_data = plate_4_QC[
+    (
+        (plate_4_QC["Metadata_heart_number"] == 7)
+        & (plate_4_QC["Metadata_treatment"] == "DMSO")
+    )
+    | (plate_4_QC["Metadata_heart_number"] == 29)
+]
+
+# Load in X and y data from dataset
+X, y = get_X_y_data(df=holdout_QC_data, label="Metadata_cell_type")
+
+# Assign y classes to correct binary using label encoder results
+y_binary_QC = le.transform(y)
+
+# Predict probabilities for the positive class
+y_probs_modelQC = QC_model.predict_proba(X)[:, 1]
+
+
+# ## Calculate the ROC AUC score from applying the respective holdout datasets to each model
+
+# In[6]:
+
+
+# Calculate ROC AUC
+aucNoQC = roc_auc_score(y_binary_no_QC, y_probs_modelNoQC)
+aucQC = roc_auc_score(y_binary_QC, y_probs_modelQC)
+
+print(f"AUC Model 1: {aucNoQC}")
+print(f"AUC Model 2: {aucQC}")
+
+
+# ## Apply bootstrap ROC function and perform a t-test to determine significant difference between distributions
+
+# In[7]:
+
+
+# No QC model
+scores_model1 = bootstrap_roc_auc(y_binary_no_QC, y_probs_modelNoQC)
+
+# QC model
+scores_model2 = bootstrap_roc_auc(y_binary_QC, y_probs_modelQC)
+
+# Compare distributions
+t_stat, p_value = ttest_ind(scores_model1, scores_model2)
+print(f"T-statistic: {t_stat}, P-value: {p_value}")
+
+
+# In[8]:
+
+
+print(f"Mean ROC AUC for Model No-QC: {np.mean(scores_model1)}")
+print(f"Mean ROC AUC for Model QC: {np.mean(scores_model2)}")
+
+
+# ## Create histogram plot with the AUC ROC bootstrap distributions
+
+# In[9]:
+
+
+# Define darker colors for the mean lines
+darker_steelblue = mcolors.to_rgba("steelblue", 0.8)
+darker_coral = mcolors.to_rgba("coral", 0.8)
+
+plt.hist(scores_model1, bins=50, alpha=0.5, label="Model No-QC", color="steelblue")
+plt.hist(scores_model2, bins=50, alpha=0.5, label="Model QC", color="coral")
+
+# Add vertical lines for the means
+plt.axvline(
+    np.mean(scores_model1),
+    color=darker_steelblue,
+    linestyle="dashed",
+    linewidth=2,
+)
+plt.axvline(
+    np.mean(scores_model2),
+    color=darker_coral,
+    linestyle="dashed",
+    linewidth=2,
+)
+
+plt.legend(loc="upper left", fontsize=10)
+plt.xlabel("ROC AUC Score", fontsize=12)
+plt.ylabel("Frequency", fontsize=12)
+plt.title("Bootstrap ROC AUC Distributions", fontsize=14, fontweight="bold")
+plt.grid(True, linestyle="--", alpha=0.6)
+plt.tight_layout()
+
+# save figure
+plt.savefig("./figures/bootstrap_plot_holdout_each_model.png", dpi=600)
```

Comment:
is it possible to be a bit more specific about what this figure represents? What does "holdout_each_model" mean?

Reply:
Updated these now!